[EDNA-70] Implement Dashboard API

gromakovsky opened the pull request:
## Description

This PR implements all endpoints from `DashboardEndpoints`.

## Related issue(s)

Resolves https://issues.serokell.io/issue/EDNA-70 and https://issues.serokell.io/issue/EDNA-32

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

gromakovsky committed:
Add Edna.Dashboard.DB.Schema

Problem: we tend to split DB schema into multiple modules, but some
tables are still in the main Edna.DB.Schema which is supposed to only
group existing tables from other components.

Solution: create Edna.Dashboard.DB.Schema module and move tables there.

gromakovsky committed:
Store new data about sub-experiments

Problem: there is new functionality that was recently added to Edna
design. Now sub-experiments have names and one of them is the primary
one. Currently we don't store this data in DB.

Solution: add a new field to the `sub_experiment` table and its
Haskell ADT.
Also add a new table that bounds an experiment with its primary
sub-experiment. We could also put this data into `experiment` table,
but then we would have circular references which are not convenient to
deal with. So I think this solution is better.

gromakovsky committed:
Add isEnabled flag to MeasurementResp

Problem: we return points on plot as a list of MeasurementResp where
each one is a pair of numbers. But we should also distinguish enabled
and disabled points somehow. Both groups of points should be shown
somehow, but they should be shown differently.

Solution: add a boolean flag to MeasurementResp.

gromakovsky committed:
Implement Dashboard getters

Problem: DashboardEndpoints are not implemented.

Solution: we start with implementing getter endpoints.
Primitive DB queries are defined in DB.Query, Service layer
does some additional logic and conversion.
There are some changes in the API:
1. Methodology ID inside `ExperimentResp` is now optional because
for some experiments there may be no methodology (if it's deleted).
2. `SubExperimentResp` doesn't contain `isDefault` flag anymore.
ID of the primary sub-experiment is returned as part of
`ExperimentResp` instead.

gromakovsky committed:
Pass --hint to hlint on CI

Problem: we have .hlint.yaml file in the repo, but it's ignored when
we run hlint on CI.
Solution: pass `--hint` argument with this file.

gromakovsky committed:
Implement remaining Dashboard endpoints

Problem: there are 4 remaining Dashboard endpoints with
stub implementation.

Solution: implement them.
DB schema was updated a bit: now all `removed_measurement` entries will
be removed when the corresponding sub-experiment is removed.
The word `default` was changed to `primary` in API.

gromakovsky committed:
Add Params4PL type and return it from RPC

Problem: in order to draw plots we need to know 4PL curve parameters,
because the curve should be built based on 4PL formula. But currently
we return only one parameter (IC50).

Solution:
1. Add a module for future analysis implmentation (that will just call
Python) with `Params4PL` type.
2. Replace `IC50 :: Double` in API types with `Params4PL`.

gromakovsky committed:
Add tests for Dashboard service

Problem: there are no tests for Dashboard functionality.

Solution: add DashboardSpec module with simple unit tests for
`Edna.Dashboard.Service`. Property tests may be added later separately.

sashasashasasha151 commented on backend/sql/drop.sql:
```diff
@@ -6,5 +6,6 @@ drop table if exists experiment_file cascade;
 drop table if exists measurement cascade;
 drop table if exists analysis_method cascade;
-drop table if exists subexperiment cascade;
+drop table if exists sub_experiment cascade;
```
Does anyone use this `drop.sql` and what for? (Just a question, I am ok with this file)

gromakovsky replied:
I often do `\i sql/drop.sql; \i sql/init.sql;` in `psql`, so yes. One day someone will simplify it. Maybe I'll even do it in this PR :)

sashasashasasha151 replied:
Why don't you use setting in config?

gromakovsky replied:
I use setting in config when I need to launch `edna-server`, but sometimes I want to do something with DB directly in which case I do `\i` in `psql`.

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -0,0 +1,154 @@
+makePrimarySubExperiment (SqlId subExpId) =
+  runUpdate' $ update (esPrimarySubExperiment ednaSchema)
+  (\pse -> pseSubExperimentId pse <-. val_ subExpId)
+  (\pse -> pseExperimentId pse ==. subquery_ (do
+    subExperiment <- all_ (esSubExperiment ednaSchema)
+    guard_ (seSubExperimentId subExperiment ==. val_ (SqlSerial subExpId))
+    return (seExperimentId subExperiment)
+    ))
```
👍

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -0,0 +1,154 @@
+
+-- | Delete a sub-experiment with given ID along with all @removed_measurement@
+-- entries for this sub-experiment.
+deleteSubExperiment :: SubExperimentId -> Edna ()
+deleteSubExperiment (SqlId subExpId) =
+  runDelete' $ delete (esSubExperiment ednaSchema) $
+  \se -> seSubExperimentId se ==. val_ (SqlSerial subExpId)
```
I thought that we discussed it and if sub experiment is primary we will not delete it 🤔

gromakovsky replied:
Yes, but currently it's checked by a separate SQL query in `Service` module (there we do `isPrimarySubExperiment` check). I decided to postpone it to [EDNA-76](https://issues.serokell.io/issue/edna-76). Though, if I manage to implement it quickly, I'll do it here soon.

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -0,0 +1,154 @@
+    primarySubExp <- all_ $ esPrimarySubExperiment ednaSchema
+    guard_ (pseExperimentId primarySubExp ==. cast_ (eExperimentId experiment) int)
+
+    subExperiment <- leftJoin_ (all_ $ esSubExperiment ednaSchema) $
+      \subExp -> cast_ (seExperimentId subExp) int ==. eExperimentId experiment
```
Why not just guard?

gromakovsky replied:
Do you mean inner join instead of left? Currently experiment always has at least one sub-experiment, so we can use inner join. My reasoning was that maybe one day we'll change something and it will be possible for an experiment to have no sub-experiments, in which case left join will be needed here.

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -0,0 +1,154 @@
+      [((ExperimentRec, Word32, Maybe Word32, LocalTime, Word32), [SqlSerial Word32])]
+    groupSubExps =
+      map (\ne -> (fst (head ne), mapMaybe snd (toList ne))) .
+      NE.groupBy (\a b -> a ^. _1 . _1 == b ^. _1 . _1)
```
As I remember `groupBy` works only on sorted lists, so maybe add ordered by to query?

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -0,0 +1,154 @@
+
+-- | Get data about all experiments using 3 optional filters: by project ID,
+-- compound ID and target ID.
+getExperiments :: Maybe ProjectId -> Maybe CompoundId -> Maybe TargetId ->
+  Edna [(ExperimentId, ExperimentResp)]
+getExperiments mProj mComp mTarget =
+  fmap (map convert . groupSubExps) $ runSelectReturningList' $ select $ do
+    experiment <- all_ $ esExperiment ednaSchema
+
+    experimentFile <- all_ $ esExperimentFile ednaSchema
+    guard_ (eExperimentFileId experiment ==. cast_ (efExperimentFileId experimentFile) int)
+
+    primarySubExp <- all_ $ esPrimarySubExperiment ednaSchema
+    guard_ (pseExperimentId primarySubExp ==. cast_ (eExperimentId experiment) int)
+
+    subExperiment <- leftJoin_ (all_ $ esSubExperiment ednaSchema) $
+      \subExp -> cast_ (seExperimentId subExp) int ==. eExperimentId experiment
+
+    whenJust mProj $ \(SqlId projId) ->
+      guard_ (efProjectId experimentFile ==. val_ projId)
+    whenJust mComp $ \(SqlId compId) ->
+      guard_ (eCompoundId experiment ==. val_ compId)
+    whenJust mTarget $ \(SqlId targetId) ->
+      guard_ (eTargetId experiment ==. val_ targetId)
```
You can leave it this way, but looks like using one filter is easier and more understandable

gromakovsky replied:
What do you mean? `Maybe (ProjectId, TargetId, CompoundId)`? We need 3 different filters due to our UI/UX.

sashasashasasha151 replied:
You can write smth like 

```haskell
filter_ (\(yourLarge tuple) -> eExperimentFileId experiment ==. cast_ (efExperimentFileId experimentFile) int &&.
pseExperimentId primarySubExp ==. cast_ (eExperimentId experiment) int &&
bool (val_ True) (efProjectId experimentFile ==. val_ projId) isJust mProj &&.
bool (val_ True) (eCompoundId experiment ==. val_ compId) isJust mComp &&.
bool (val_ True) (eTargetId experiment ==. val_ targetId) isJust mTarget &&. )
```

But for real I am ok with both solutions, just what you prefer

sashasashasasha151 commented on backend/src/Edna/Web/Server.hs:
```diff
@@ -73,6 +74,7 @@ ednaToHandler ctx action =
   `catch` throwPurely @UploadError
   `catch` throwPurely @UploadApiError
   `catch` throwPurely @LibraryError
+  `catch` throwPurely @DashboardError
```
Why do we need to list all errors here?

gromakovsky replied:
Because if you throw an exception, Servant doesn't know how to handle it and treats it as an internal server error, returning the corresponding HTTP code. We want to map all these errors to something different from 500, so we need to catch them and let Servant know what to do with them. And all errors not listed here will be treated as 500.

sashasashasasha151 replied:
I think I know the solution how to not write it here but will check later

sashasashasasha151 commented on backend/src/Edna/Dashboard/Service.hs:
```diff
@@ -0,0 +1,130 @@
+makePrimarySubExperiment :: SubExperimentId ->
+  Edna (WithId 'SubExperimentId SubExperimentResp)
+makePrimarySubExperiment subExpId = do
+  logMessage $ fmt $ "Making sub-experiment " +| subExpId |+ " primary"
```
We already log request with responses, looks like no need to duplicate it

gromakovsky replied:
1. We log them only if logging is `Dev`. And if logging mode is `Prod`, my idea was to log some important actions briefly. And here I tentatively treat only mutating actions (i. e. not getters) as _important_. So in `Prod` mode you will see state changes, but won't see all requests, so the log won't be too big.
2. In `Dev` mode (which I treat as `Debug` mode) this duplication may be not bad because sometimes weird things happen, e. g. an endpoint is called, but no state is changed, and these messages help you understand whether you reached `Q.makePrimarySubExperiment`. Also they are easier to read than full requests because they are written in human language.

That said, I see your point too, so if I didn't convince you that these messages are ok, I'll delete them.

sashasashasasha151 replied:
>So in Prod mode you will see state changes, but won't see all requests, so the log won't be too big.

As I already wrote, on prod it is also very useful to have full logs

>I'll delete them

Lets leave them to merge faster

gromakovsky replied:
> As I already wrote, on prod it is also very useful to have full logs

Yes, actually, it's set to Dev in both staging and demo for now: https://github.com/serokell/gemini-infra/pull/10/files
I don't have a real use case for Prod for now, the time will show whether it's useful or not.

sashasashasasha151 commented on backend/src/Edna/Dashboard/Service.hs:
```diff
@@ -0,0 +1,130 @@
+getExperiments mProj mComp mTarget = do
+  pairs <- Q.getExperiments mProj mComp mTarget
+  meanIC50 <- case (mComp, mTarget) of
+    (Just _, Just _) -> Just <$> computeMeanIC50 (map snd pairs)
```
What mean `MeanIC50` for scientists?

gromakovsky replied:
AFAIU the person who decides whether a compound is good or not is interested in IC50 value for `(compound, target that is supposed to be attacked by this compound)`. If they select this compound and target in dashboard, they will see potentially many experiments with this pair and values may be different in these experiments (AFAIU in reality values should be similar, but they can't be absolutely equal). But it's hard to make a decision when you see 10 values, so we also provide average value.

I might be wrong, but I vaguely remember "mean IC50" being mentioned on some meetings, and that's how it appeared in designs.

gromakovsky committed:
Simplify drop.sql

Problem: currently we list all tables in drop.sql files and it's
tedious to maintain this list.

Solution: simplify this file to drop schema and create it again,
we are already doing the same thing when `enable-with-drop` mode
is used.
